lua/apisix/plugins/zipkin.lua

```diff
@@ -0,0 +1,129 @@
+    },
+    required = {"endpoint"}
+}
+
+
+local _M = {
+    version = 0.1,
+    priority = -1000, -- last running plugin
+    name = plugin_name,
+    schema = schema,
+}
+
+
+function _M.check_schema(conf)
+    local ok, err = core.schema.check(schema, conf)
+
+    if not ok then
+        return false, err
+    end
+
+    return true
+end
+
+
+local function create_tracer(conf)
+	local tracer = new_tracer(new_reporter(conf), new_random_sampler(conf))
+	tracer:register_injector("http_headers", zipkin_codec.new_injector())
+	tracer:register_extractor("http_headers", zipkin_codec.new_extractor())
+	return tracer
+end
+
+local function report2endpoint(premature, reporter)
+	if premature then
+		return
+	end
+
+	local ok, err = reporter:flush()
+	if not ok then
+		core.log.error("reporter flush ", err)
+		return
+	end
+end
+
+
+function _M.rewrite(conf, ctx)
+    local tracer = core.lrucache.plugin_ctx(plugin_name, ctx,
+                                                 create_tracer, conf)
+
+    local wire_context = tracer:extract("http_headers", core.request.headers(ctx))
+
+    local start_timestamp = ngx.req.start_time()
+    local request_span = tracer:start_span("kong.request", {
+        child_of = wire_context,
+        start_timestamp = start_timestamp,
+        tags = {
+            component = "apisix",
+            ["span.kind"] = "server",
+            ["http.method"] = ctx.var.method,
+            ["http.url"] = ctx.var.request_uri,
+            ["peer.ipv4"] = core.request.get_remote_client_ip(ctx), -- TODO: support ipv6
+            ["peer.port"] = core.request.get_remote_client_port(ctx),
+        }
+    })
+
+    local rewrite_span = ctx.opentracing.rewrite_span:start_child_span(
+                                            "apisix.rewrite", start_timestamp)
+
+    ctx.opentracing = {
+        tracer = tracer,
+        wire_context = wire_context,
+        request_span = request_span,
+        rewrite_span = rewrite_span,
+        access_span = nil,
+        proxy_span = nil,
+        header_filter_span = nil,
+        header_filter_finished = false,
+        body_filter_span = nil,
+    }
+
+    ctx.REWRITE_END_TIME = tracer:time()
+    ctx.opentracing.rewrite_span:finish(ctx.REWRITE_END_TIME)
+end
+
+function _M.access(conf, ctx)
+    local opentracing = ctx.opentracing
+
+    opentracing.proxy_span = ctx.opentracing.access_span:start_child_span(
+            "apisix.access", ctx.KONG_ACCESS_START / 1000)
+
+		-- Want to send headers to upstream
+		local outgoing_headers = {}
+		opentracing.tracer:inject(opentracing.proxy_span, "http_headers", outgoing_headers)
+		local set_header = kong.service.request.set_header
+		for k, v in pairs(outgoing_headers) do
+			set_header(k, v)
+        end
+
+end
+
+
+function _M.log(conf, ctx)
+	local tracer = ctx.opentracing.tracer
+	local reporter = tracer.reporter
+	local ok, err = ngx.timer.at(0, report2endpoint, reporter)
```

Comment:
that a bad way to create a timer for each request. 
we can add a todo comment first, will fix it later.